## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [511]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [512]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [513]:
# Use the describe function to gather some basic statistics
df.describe


<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [514]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [515]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [516]:
# For the category with the most entries,
# which subcategory had the most entries?
df[["category","subcategory"]].value_counts().head(1)




category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [517]:
# Which five clients had the most entries in the data?
df[["client_id"]].value_counts().head()


client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64

In [518]:
# Store the client ids of those top 5 clients in a list.
top5 = []
clients = df["client_id"].value_counts().head()
#clients = df["client_id"].value_counts().reset_index().rename(columns={"index": "client-id"}).head()
top5 = clients.index.values[0:5].tolist()
print(top5)


[33615, 66037, 46820, 38378, 24741]


In [519]:
# How many total units (the qty column) did the
# client with the most entries order order?
orders_df = df.loc[(df["client_id"] == 33615)]
sumdorders_df = orders_df.sum()
print(f'{sumdorders_df.iloc[16]}')



64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [520]:
df2 = pd.read_csv('Resources/client_dataset.csv')

In [521]:
df2.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [522]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df2['line_subtotal'] = (df2['unit_price'] * df2['qty'])
df2.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [523]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def pricer(x):
     if x>50:
        return 7.00
     else:
        return 10.00  
     
x = (df2['unit_weight'] * df2['qty'])
df2['shipping_price'] = x.apply(pricer) * (df2['unit_weight'] * df2['qty'])
df2.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  \
0     1096.80     762.71         7.50  105            1      115164.00   
1       24.95      15.09         1.49   21            0         523.95   
2       13.52       7.86         1.68   39            6         527.28   
3       36.42      24.85         1.23   29            3        1056.18   
4      195.10     108.17        46.43   20            1        3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [524]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df2['line_price'] = (df2['line_subtotal'] + df2['shipping_price']) * .0925 + (df2['line_subtotal'] + df2['shipping_price'])
df2.head().round(2)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  shipping_price  \
0     762.71         7.50  105            1      115164.00         5512.50   
1      15.09         1.49   21            0         523.95          312.90   
2       7.86         1.68   39            6         527.28          458.64   
3      24.85         1.23   29            3        1056.18          356.70   
4     108.17        46.43   20            1        3902.00         6500.20   

   line_price  
0   131839.08  
1      914.26  
2     1077.12  
3     1543.57  
4    11364.40  

[5 rows x 21 columns]

In [525]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df2['line_cost'] = (df2['unit_cost'] * df2['qty']) +  df2['shipping_price']
df2.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price  \
0         7.50  105            1      115164.00         5512.50   
1         1.49   21            0         523.95          312.90   
2         1.68   39            6         527.28          458.64   
3         1.23   29            3        1056.18          356.70   
4        46.43   20            1        3902.00         6500.20   

      line_price  line_cost  
0  131839.076250   85597.05  
1     914.258625     629.79  
2    1077.117600     765.18  
3    1543.571400    1077.35  
4   11364.403500    8663.60  

[5 rows x 22 columns]

In [526]:
# Create a column for the profit of
# each line using line cost and line price
df2['line_profit'] = (df2['line_price'] - df2['line_cost'])
df2.head().round(2)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  shipping_price  line_price  line_cost  \
0            1      115164.00         5512.50   131839.08   85597.05   
1            0         523.95          312.90      914.26     629.79   
2            6         527.28          458.64     1077.12     765.18   
3            3        1056.18          356.70     1543.57    1077.35   
4            1        3902.00         6500.20    11364.40    8663.60   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [527]:
# Check your work using the totals above
orders_df2 = df2.loc[(df["order_id"] == 2742071)]
sumdorders_df2 = orders_df2.sum()
print(f'Order 2742071 total: ${sumdorders_df2.iloc[20]:.3f}')
orders_df2 = df2.loc[(df["order_id"] == 2173913)]
sumdorders_df2 = orders_df2.sum()
print(f'Order 2173913 total: ${sumdorders_df2.iloc[20]:.3f}')
orders_df2 = df2.loc[(df["order_id"] == 6128929)]
sumdorders_df2 = orders_df2.sum()
print(f'Order 6128929 total: ${sumdorders_df2.iloc[20]:.3f}')


Order 2742071 total: $152811.897
Order 2173913 total: $162388.719
Order 6128929 total: $923441.240


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [528]:
top5_df = pd.Series(top5)
#print(f'{top5_df}')
orders_df2 = df2.loc[(df2["client_id"].isin(top5_df))] 



In [529]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
def order_sum(x):
    return x.sum()


orders_df3 = orders_df2.groupby(["client_id"]).\
            apply(lambda x: pd.Series({"Total_Sum_of_Orders": order_sum(x["line_price"])}))
orders_df3["Total_Sum_of_Orders"] = orders_df3["Total_Sum_of_Orders"].map('${:,.2f}'.format)
orders_df3


Total_Sum_of_Orders
client_id                    
24741          $82,268,891.98
33615           $8,377,308.57
38378          $12,906,550.88
46820           $9,743,794.32
66037          $10,259,514.80

In [530]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
def order_sum(x):
    return x.sum()


orders_df4 = orders_df2.groupby(["client_id"]).\
            apply(lambda x: pd.Series({"qty": order_sum(x["qty"]),                            
                                       "shipping_price": order_sum(x["shipping_price"]),
                                       "line_price": order_sum(x["line_price"]),
                                       "line_cost": order_sum(x["line_cost"]),
                                       "line_profit": order_sum(x["line_profit"])}))
orders_df4["qty"] = orders_df4["qty"].map('{:,.0f}'.format)
orders_df4["shipping_price"] = orders_df4["shipping_price"].map('${:,.2f}'.format)
orders_df4["line_price"] = orders_df4["line_price"].map('${:,.2f}'.format)
orders_df4["line_cost"] = orders_df4["line_cost"].map('${:,.2f}'.format)
orders_df4["line_profit"] = orders_df4["line_profit"].map('${:,.2f}'.format)
orders_df5 = orders_df4.sort_values(by=["line_profit",],ascending=False)
orders_df5



qty shipping_price      line_price       line_cost  \
client_id                                                           
24741      239,862  $5,126,448.37  $82,268,891.98  $45,688,899.71   
38378       73,667  $3,429,455.40  $12,906,550.88   $9,634,720.98   
66037       43,018  $1,395,151.85  $10,259,514.80   $7,004,482.98   
46820       75,768  $1,601,448.84   $9,743,794.32   $7,007,191.64   
33615       64,313  $1,828,984.89   $8,377,308.57   $6,175,313.91   

              line_profit  
client_id                  
24741      $36,579,992.27  
38378       $3,271,829.90  
66037       $3,255,031.82  
46820       $2,736,602.68  
33615       $2,201,994.66

In [531]:
# Sort the updated data by "Total Profit" form highest to lowest


def order_sum(x):
    return x.sum()


orders_df8 = orders_df2.groupby(["client_id"]).\
            apply(lambda x: pd.Series({"Units": order_sum(x["qty"]),                            
                                       "shipping_price": order_sum(x["shipping_price"]/1000000),
                                       "line_price": order_sum(x["line_price"]/1000000),
                                       "line_cost": order_sum(x["line_cost"]/1000000),
                                       "line_profit": order_sum(x["line_profit"]/1000000)}))
orders_df9 = orders_df8.rename(columns={"shipping_price": "Shipping", "line_price": "Total Revenue", \
                                        "line_cost": "Total Cost", \
                                        "line_profit": "Total Profit"}).map('${:,.2f}M'.format)
orders_df10 = orders_df9.sort_values(by=["Total Profit",],ascending=False)
orders_df10

Units Shipping Total Revenue Total Cost Total Profit
client_id                                                             
24741      $239,862.00M   $5.13M       $82.27M    $45.69M      $36.58M
38378       $73,667.00M   $3.43M       $12.91M     $9.63M       $3.27M
66037       $43,018.00M   $1.40M       $10.26M     $7.00M       $3.26M
46820       $75,768.00M   $1.60M        $9.74M     $7.01M       $2.74M
33615       $64,313.00M   $1.83M        $8.38M     $6.18M       $2.20M